content/realtime/channel-params.textile

```diff
@@ -0,0 +1,91 @@
+
+In an SSE connection, it is also possible to specify channel parameters as a query string in the connection URL, instead of as a qualifier on an individual channel name. In this case, the given channel parameters apply to all channel attachments associated with that connection.
+
+h2(#rewind). rewind parameter
+
+The @rewind@ channel parameter relates to the initial attachment of a connection to a channel, and expresses the intent to attach to the channel at a position, or a point in time, in the past (that is, effectively "rewinding" the channel for the purposes of the present attachment).
+
+A @rewind@ parameter can express a channel position in terms of a number of messages, or a time interval.
+
+A @rewind@ value that is simply a number @n@ (eg @rewind=1@) is a request to attach to the channel at a position @n@ messages before the present position. If that attachment is successful, and one or more messages exist on the channel prior to the present position, then those messages will be delivered to the subscriber immediately after the attachment has completed, and before any subsequent messages that arise in real time.
+
+If fewer than the requested number of messages exists on the channel (including the case that there are no prior messages), then the available messages are sent; this does not constitute an error.
+
+A @rewind@ value can also be a string that is a time interval specifier. Supported specifier values express an integral number of seconds (eg @15s@), minutes (eg @5m@) or hours (eg @4h@). If that attachment is successful, and one or more messages exist on the channel in the given time interval prior to the present time, then those messages will be delivered to the subscriber immediately after the attachment has completed, and before any subsequent messages that arise in real time.
+
+If fewer than the requested number of messages exists on the channel in that interval (including the case that there are no messages), then the available messages are sent; this does not constitute an error.
+
+If the number of messages within the specified interval is greater than the attachment @limit@ (which defaults to 100), then only the most recent messages up to that limit are sent. The attachment succeeds, but truncation of the message backlog is indicated as a non-fatal error in the attachment response.
+
+It is also possible to specify a different limit via a @limit@ channel parameter. So, for example, to request up to 10 messages, in a window 5m before the present time, specify a channel parameter string of @[?rewind=5m&limit=10]@.
+
+By default, a maximum of two minutes of channel history is available when attaching. This means that a rewind time specifier of greater than two minutes will only be able to rewind by two minutes. If a channel has persistence enabled, then it is possible to rewind back in time by up to the persistence TTL on the channel. 
```

Comment:
> If a channel has persistence enabled, then it is possible to rewind back in time by up to the persistence TTL on the channel. 

Not currently it isn't.

The rewind-reaching-into-persisted-history work was done with the goal of supporting rewind=1 even if it has to reach into persistence; arbitrary time-based rewind beyond the two-minute window is currently neither tested nor enabled. It might work without many further changes, but if we want it to, we would still need to do the work to implement that. (And even with rewind=1 it isn't currently enabled by default -- we still need to decide when/how to do that)

Reply:
>  It might work without many further changes, but if we want it to, we would still need to do the work to implement that. (And even with rewind=1 it isn't currently enabled by default -- we still need to decide when/how to do that)

That seems like an arbitrary limitation.  Why would we limit history to two minute queries for persisted history?  That makes absolutely no sense IMHO.

Reply:
> That seems like an arbitrary limitation. Why would we limit history to two minute queries for persisted history? That makes absolutely no sense IMHO.

Rewind is a parameter that requests a specific attach point, like resume; that's all it's ever been. We added a kludge where rewind=1 can reach back into persisted history to support data-object use-cases, but we don't currently have a feature where resuming (whether from a channelSerial, a timestamp, or just a limit) on a persisted channel can go back to an arbitrary point in time.

We have previously considered allowing that -- arbitrary resuming on persisted channels, no more two minute window, yay. We even came up with some concrete plans for how to do so (e.g. switching timelinequery to index by the channelTimeserial rather than the frontend timestamp). But we decided against that, in favour of putting our effort into replacing both live and persisted history with NATS/Liftbridge. Did you change your mind about that?

Reply:
> we don't currently support resuming (whether from a channelSerial, a timestamp, or just a limit) back to an arbitrary point in time on persisted channels.

I understood that we were supporting that for a numeric limit (ie `rewind=10`), although there is a disparity about what the number means (ie messages vs protocol messages). Rewind into persisted history via a time specifier I agree isn't supportable without a change to `TimelineQuery` to apply time bounds based on channel timeserial (which is possible, but not highest priority).

Reply:
>  Did you change your mind about that?

I was not made aware of those issues.